src/widgets/scrollbar.rs

```diff
@@ -446,198 +449,177 @@ impl<'a> Scrollbar<'a> {
-        //                    state.viewport_content_length;
-        // let thumb_end = thumb_size + thumb_start;
-        // (thumb_start, thumb_end)
-
-        let (track_start, track_end) = track_start_end;
-
-        let viewport_content_length = if state.viewport_content_length == 0 {
-            (track_end - track_start) as usize
-        } else {
-            state.viewport_content_length
+    /// ```plain
+    ///         ┌──────────── thumb_len
+    ///       vvvvv
+    ///    ═══█████═══════
+    ///    ^^^     ^^^^^^^
+    ///     │         └────── track_end_len
+    ///     │
+    ///     └──────────────── track_start_len
+    /// ```
+    fn get_track_lens(&self, area: Rect, state: &mut ScrollbarState) -> (usize, usize, usize) {
+        let (mut track_start, mut track_end) = match self.orientation {
+            ScrollbarOrientation::VerticalRight => (area.y, (area.y + area.height)),
+            ScrollbarOrientation::VerticalLeft => (area.y, (area.y + area.height)),
+            ScrollbarOrientation::HorizontalBottom => (area.x, (area.x + area.width)),
+            ScrollbarOrientation::HorizontalTop => (area.x, (area.x + area.width)),
         };
+        // if scrollbar has begin and end symbols:
+        //
+        // <═══█████═══════>
+        //
+        // then increment and decrement track_start and track_end respectively
+        if let Some(s) = self.begin_symbol {
+            track_start = track_start.saturating_add(s.width() as u16);
+        };
+        if let Some(s) = self.end_symbol {
+            track_end = track_end.saturating_sub(s.width() as u16);
+        };
+        let track_len = track_end.saturating_sub(track_start) as f64;
 
-        let scroll_position_ratio = (state.position as f64 / state.content_length as f64).min(1.0);
-
-        let thumb_size = (((viewport_content_length as f64 / state.content_length as f64)
-            * (track_end - track_start) as f64)
-            .round() as u16)
-            .max(1);
+        let viewport_len = self.get_viewport_len(area) as f64;
 
-        let track_size = (track_end - track_start).saturating_sub(thumb_size);
+        let content_length = state.content_length as f64;
+        // if user passes in position > content_length, we shouldn't panic
+        // this will prevent rendering outside of available area
+        let position = state.position.min(state.content_length - 1) as f64;
 
-        let thumb_start = track_start + (scroll_position_ratio * track_size as f64).round() as u16;
+        // vscode style scrolling behavior
+        let scrollable_content_len = content_length + viewport_len - 1.0;
+        let thumb_start = position * track_len / scrollable_content_len;
+        let thumb_end = (position + viewport_len) * track_len / scrollable_content_len;
 
-        let thumb_end = thumb_start + thumb_size;
+        // round() as usize gives closest int, as opposed to `floor` or `ceil`
```

Comment:
Make a note that this intentionally rounds the positions and not the sizes, as rounding the sizes leads to subtle off by 1 errors. e.g. 6.4 (position) +3.4 (length) = 9.8 which rounds to 10, but 6 (rounded position) + 3 (rounded length) = 9.